packages/accounts/src/light-account/utils.ts

```diff
@@ -15,44 +16,100 @@ import {
   polygonAmoy,
   polygonMumbai,
   sepolia,
+  type AccountVersionAddresses,
+  type EntryPointVersion,
+  type OneOf,
   type SmartContractAccount,
 } from "@alchemy/aa-core";
 import { fromHex, type Address, type Chain } from "viem";
 
-export type LightAccountVersion =
-  /** @deprecated This version does not support 1271 signature validation */
-  | "v1.0.1"
-  /** @deprecated This version has a known issue with 1271 validation, it is recommended to use v1.1.0 */
-  | "v1.0.2"
-  | "v1.1.0";
-
-export const LightAccountVersions: Record<
-  LightAccountVersion,
-  {
-    factoryAddress: Address;
-    implAddress: Address;
-  }
+export type LightAccountType = keyof ILightAccountVersions;
+
+export type LightAccountVersionDef<
+  TEntryPointVersion extends EntryPointVersion
 > = {
+  factoryAddress: Address;
+  implAddress: Address;
+  entryPointVersion: OneOf<TEntryPointVersion, EntryPointVersion>;
+};
+
+export interface ILightAccountVersions {
```

Comment:
I'm not a huge fan of this setup where we use `keyof ` to represent the versions instead of the union type of strings before. We do this with `EntryPointRegistry` as well and it's doesn't seem idiomatic, as I don't see this pattern used anywhere else. @dphilipson would love your thoughts here as well

I'd rather we leverage the pattern that we had before and not try to merge MultiOwner and single owner LA into the same logic. Keeping them separate will be easier imo and we can share logic between them with helper functions. A good example of this is multisig MA vs multiowner MA.

One thing about using `keyof` is that it now masks the underlying accepted values from developers. If I hover over a type that uses this, I don't see `"v1.0.1" | ...` I just see, `keyof ...` as a developer I need to now go and find what that thing I'm taking `keyof` is if I want to understand the accepted values.

Separately having hard coded types like `LightAccountConfig<"v1.0.1", "0.6.0">` adds to the maintenance cost and is something that can easily be done with typescript over the generic anyways. You can restrict the versions where it matters (ie. LightAccount and LightAccountParams) with something like this:

```ts
LightAccount<
  TSigner extends SmartAccountSigner = SmartAccountSigner,
  TVersion extends LightAccountVersion = LightAccountVersion
  TEntryPointVersion extends EntryPointVersion = TVersion extends `v1${string}` ? "0.6.0" : "0.7.0"
>

// similar thing for LightAccountParams
```

Doing this means that supporting a new version is handled by just adding to the union string type and recording the new `impl` and `factoryAddresses`.

With the registry flow, you need to update the Registry type as well as the versions map. While that's still just 2 updates, the updates requires a bit more than in the above (this is more of a nit than anything else).

**The idiomatic typescript is more the feedback here:**
Consider `viem`'s `Client` type. It's generic on a lot of other things, but assume for a second it's only generic on `Transport` which can take multiple shapes, but let's stick to `HttpTransport | CustomTransport`. 

They just have their `Client` definition generic on `Transport` and when `HttpTransport` is used or `CustomTransport` is used the rest of the typing just follows that. We don't also maintain a `ClientRegistry` like this

```ts
type ClientRegistry = {
  HttpTransport: Client<HttpTransport>
  CustomTransport: Client<CustomTransport>
}
```

The reason is that it's an added layer of abstraction that becomes more restrictive and is already handled by typescript. Example usage:
```ts
function getClient<TTransport extends Transport>(transport: TTransport): ClientRegistery[TTransport];

// is equivalent to

function getClient<TTransport extends Transport>(transport: TTransport): Client<TTransport>;
```

Reply:
@dphilipson thoughts on this? @moldy530 here are the benefits of using typescript interface/type and using keyof to define versions (string). 
- we can enforce that keys do exist in the configured version definition maps and enforce fully defined versions are used by the users.
- We can distinguish between the value type of the keyed versions by doing `Interface[key]` and that will specifically give me the type of version def typed with the key.

Reply:
For the entry point versions, it looks to me like you'd be able to do
```ts
export type EntryPointVersion = "0.6.0" | "0.7.0";
export type EntryPointRegistryBase<T> = Record<EntryPointVersion, T>;
```
Is there a reason not to do that?

The light account version case is a little trickier to write without the `keyof`. I might need to think about that one a bit. I definitely agree with Moldy that having the version type declared as unions improves error messages, but the setup with the registry with light account type keys is a pretty good one.

Reply:
Also thinking about this a bit more, I'm not positive I like having the same function give you a LightAccount or MultiOwnerLightAccount differing just on one optional parameter. This seems like it could be confusing or error-prone for MultiOwnerLightAccount users, who can very easily forget to set that parameter, have everything typecheck, but have all their calls fail because they're using the wrong account type.

I feel the same way about defaulting to a certain version, actually- it would be really easy for a 2.0 user to forget to pass in the optional version parameter and then have everything cryptically fail. I wonder if it might make sense to deprecate the function where the light account version is optional and always require users to pass in the light account version.

Reply:
> I'm not positive I like having the same function give you a LightAccount or MultiOwnerLightAccount differing just on one optional parameter

yea this is an area where I could see devs getting confused. It also makes the docs slightly more complex because we have just one doc with a lot of branching sections. Would rather we be explicit for devs and just have `createLA` and `createMultiOwnerLA` 

> I feel the same way about defaulting to a certain version, actually- it would be really easy for a 2.0 user to forget to pass in the optional version parameter and then have everything cryptically fail

yea I agree we should deprecate the function that allows for optional version and force devs to pin their code to a specific version

Reply:
we have explicit createLA/createMultiOwnerLA functions. Each call createLABase respectively which calls toSmartcontractaccount. cc: @dphilipson @moldy530